src/coreclr/jit/lclmorph.cpp

```diff
@@ -175,6 +175,268 @@ void Compiler::fgSequenceLocals(Statement* stmt)
+
+        return false;
+    }
+
+    //-------------------------------------------------------------------
+    // StartBlock: Start a new block by computing incoming assertions for the
+    // block.
+    //
+    // Arguments:
+    //   block - The block
+    //
+    void StartBlock(BasicBlock* block)
+    {
+        if ((m_assertions.Height() == 0) || (block->bbPreds == nullptr) || m_comp->bbIsHandlerBeg(block))
+        {
+            m_currentAssertions = 0;
+            return;
+        }
+
+        m_currentAssertions = UINT64_MAX;
+        for (BasicBlock* pred : block->PredBlocks())
+        {
+            assert(m_comp->m_dfsTree->Contains(pred));
+            if (pred->bbPostorderNum <= block->bbPostorderNum)
+            {
+                m_currentAssertions = 0;
+                break;
+            }
+
+            m_currentAssertions &= m_outgoingAssertions[pred->bbPostorderNum];
+        }
+
+#ifdef DEBUG
+        if (m_currentAssertions != 0)
+        {
+            JITDUMP(FMT_BB " incoming assertions:\n", block->bbNum);
+            uint64_t assertions = m_currentAssertions;
+            do
+            {
+                uint32_t index = BitOperations::BitScanForward(assertions);
+                JITDUMP("  A%02u: ", index);
+                DBEXEC(VERBOSE, m_assertions.BottomRef((int)index).Print());
+                JITDUMP("\n");
+
+                assertions ^= uint64_t(1) << index;
+            } while (assertions != 0);
+        }
+#endif
+    }
+
+    //-------------------------------------------------------------------
+    // EndBlock: End a block by recording its final outgoing assertions.
+    //
+    // Arguments:
+    //     block - The block
+    //
+    void EndBlock(BasicBlock* block)
+    {
+        m_outgoingAssertions[block->bbPostorderNum] = m_currentAssertions;
+    }
+
+    //-------------------------------------------------------------------
+    // OnExposed: Mark that a local is having its address taken.
+    //
+    // Arguments:
+    //   lclNum - The local
+    //
+    void OnExposed(unsigned lclNum)
+    {
+        BitVecTraits localsTraits(m_comp->lvaCount, m_comp);
+        BitVecOps::AddElemD(&localsTraits, m_localsToExpose, lclNum);
+    }
+
+    //-------------------------------------------------------------------
+    // Record: Record an assertion about the specified local.
+    //
+    // Arguments:
+    //   dstLclNum - Destination local
+    //   dstOffs   - Offset into the local
+    //   srcLclNum - Local having its address taken
+    //   srcOffs   - Offset into the source local of the address being taken
+    //
+    void Record(unsigned dstLclNum, unsigned srcLclNum, unsigned srcOffs)
+    {
+        LocalEqualsLocalAddrAssertion assertion(dstLclNum, srcLclNum, srcOffs);
+
+        unsigned* index = m_map.LookupPointerOrAdd(assertion, UINT_MAX);
+
+        if (*index == UINT_MAX)
+        {
+            if (m_assertions.Height() >= 64)
+            {
+                return;
```

Comment:
Do you have any data on how often we drop assertions here? Might be worth noting this in the dump or something similar.

Reply:
Looks like these contexts:
```scala
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\coreclr_tests.run.windows.x64.checked.mch # 66062 : IL size 617
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\coreclr_tests.run.windows.x64.checked.mch # 66094 : IL size 617
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\coreclr_tests.run.windows.x64.checked.mch # 30938 : IL size 1066
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\benchmarks.run.windows.x64.checked.mch # 9469 : IL size 2145
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests.run.windows.x64.Release.mch # 396680 : IL size 2145
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests.run.windows.x64.Release.mch # 399807 : IL size 2145
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch # 168705 : IL size 2145
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch # 168329 : IL size 2145
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch # 168914 : IL size 2145
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch # 167285 : IL size 2145
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch # 168991 : IL size 2145
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\benchmarks.run.windows.x64.checked.mch # 7002 : IL size 2445
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch # 167037 : IL size 2445
[10:52:21]   C:\dev\dotnet\spmi\mch\227e46fa-1be3-4770-b613-4a239e7c28aa.windows.x64\libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch # 169019 : IL size 2853
```

I'll add some JITDUMP. Also, I realized this code ends up adding to the map even if we ran out of assertion space. I'll fix that as well.